src/rgw/rgw_rest.cc

```diff
@@ -801,16 +791,40 @@ void abort_early(struct req_state *s, RGWOp *op, int err_no,
        *   x-amz-error-message: The specified key does not exist.
        *   x-amz-error-detail-Key: foo
        */
-      end_header(s, op, NULL, error_content.size(), false, true);
+      end_header(s, handler, dump_more, NULL, error_content.size(), false, true);
       RESTFUL_IO(s)->send_body(error_content.c_str(), error_content.size());
     } else {
-      end_header(s, op);
+      end_header(s, handler, dump_more);
     }
     rgw_flush_formatter(s, s->formatter);
   }
   perfcounter->inc(l_rgw_failed_req);
 }
 
+void abort_early(struct req_state *s, RGWOp* op, int err_no,
+		RGWHandler* handler)
+{
+  string error_content("");
+  boost::function<void()> dump_more;
+
+  // op->error_handler is responsible for calling it's handler error_handler
+  if (op != NULL) {
+    int new_err_no;
+    new_err_no = op->error_handler(err_no, &error_content);
+    ldout(s->cct, 20) << "op->ERRORHANDLER: err_no=" << err_no
+		      << " new_err_no=" << new_err_no << dendl;
+    err_no = new_err_no;
+  } else if (handler != NULL) {
+    int new_err_no;
+    new_err_no = handler->error_handler(err_no, &error_content);
+    ldout(s->cct, 20) << "handler->ERRORHANDLER: err_no=" << err_no
+		      << " new_err_no=" << new_err_no << dendl;
+    err_no = new_err_no;
+  }
+  if (op) dump_more = op->dump_access_control_f();
+  abort_early(s, dump_more, error_content, err_no, handler);
```

Comment:
there is a check for `if (handler != NULL)` above - can you tell whether we have callers that will actually pass null?

and maybe add an `assert(handler);` before your calls to `handler->set_rgw_err()` and `handler->dump()` to make sure we aren't hitting that in teuthology